### PR TITLE
gRPC code gen to generate health and reflection stubs

### LINKED ISSUE #10214: Quarkus project compilation fails when && is present in path - e.g. 'RHEL8 && large'
Quarkus project compilation fails when && is present in path - e.g. 'RHEL8 && large'
I have Jenkins matrix job running on RHEL 8 slaves with large size and Quarkus compilation fails.

Reproducer:
```
mkdir 'RHEL8 && large'
cd 'RHEL8 && large'
git clone git@github.com:quarkusio/quarkus.git
cd quarkus
mvn clean install -DskipTests -DskipITs -DskipDocs
```

Quite interesting observation is that the build passes when only the `extensions/grpc/runtime` module is picked, but it fails when `extensions/grpc` or full build is executed
```
mvn clean install -DskipTests -DskipITs -DskipDocs -f extensions/grpc/runtime => PASS
mvn clean install -DskipTests -DskipITs -DskipDocs -f extensions/grpc => FAIL
```

Details:
```
[INFO] ----------------------< io.quarkus:quarkus-grpc >-----------------------
[INFO] Building Quarkus - gRPC - Runtime 999-SNAPSHOT                 [278/646]
[INFO] --------------------------------[ jar ]---------------------------------
...
[INFO] --- protobuf-maven-plugin:0.6.1:compile (compile) @ quarkus-grpc ---
[INFO] Building protoc plugin: quarkus-grpc-protoc-plugin
[INFO] Compiling 2 proto file(s) to /Users/rsvoboda/tmp/RHEL8 && large/quarkus/extensions/grpc/runtime/target/generated-sources/protobuf/java
[ERROR] PROTOC FAILED: Error: Could not find or load main class &&
Caused by: java.lang.ClassNotFoundException: &&
--quarkus-grpc-protoc-plugin_out: protoc-gen-quarkus-grpc-protoc-plugin: Plugin failed with status code 1.

[ERROR] /Users/rsvoboda/tmp/RHEL8 && large/quarkus/extensions/grpc/runtime/src/main/proto/health/v1/health.proto [0:0]: Error: Could not find or load main class &&
Caused by: java.lang.ClassNotFoundException: &&
--quarkus-grpc-protoc-plugin_out: protoc-gen-quarkus-grpc-protoc-plugin: Plugin failed with status code 1.

[ERROR] /Users/rsvoboda/tmp/RHEL8 && large/quarkus/extensions/grpc/runtime/src/main/proto/reflection/v1alpha/reflection.proto [0:0]: Error: Could not find or load main class &&
Caused by: java.lang.ClassNotFoundException: &&
--quarkus-grpc-protoc-plugin_out: protoc-gen-quarkus-grpc-protoc-plugin: Plugin failed with status code 1.
```

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
fixes #10214

### CODE REVIEW

**Review by cescoffier — APPROVED** [review]

File: extensions/grpc/codegen/pom.xml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>quarkus-grpc-parent</artifactId>
+    <groupId>io.quarkus</groupId>
+    <version>999-SNAPSHOT</version>
+    <relativePath>../</relativePath>
+  </parent>
+
+  <artifactId>quarkus-grpc-codegen</artifactId>
+  <name>Quarkus - gRPC - Code Gen</name>
```
Comment (gsmet) [inline extensions/grpc/codegen/pom.xml:14]: I would prefer if we could stick to the four spaces convention.

It's not consistent across all the source code but most of our pom files are this way.

File: extensions/grpc/codegen/pom.xml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      <groupId>io.quarkus</groupId>
+      <artifactId>quarkus-devtools-utilities</artifactId>
+    </dependency>
+
+    <dependency>
+      <groupId>com.google.protobuf</groupId>
+      <artifactId>protoc</artifactId>
+      <classifier>linux-aarch_64</classifier>
+      <type>exe</type>
+      <version>${protoc.version}</version>
```
Comment (gsmet) [inline extensions/grpc/codegen/pom.xml:31]: Shouldn't the version be in the BOM? Or there's a good reason to have them here?

File: extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java [hunk 3]
```diff
@@ -94,6 +95,14 @@ public boolean trigger(CodeGenContext context) throws CodeGenException {
         return false;
     }
 
+    private String escapeWhitespace(String path) {
+        if (OS.determineOS() == OS.LINUX) {
+            return path.replaceAll(" ", "\\ ");
+        } else {
+            return path;
```
Comment (gsmet) [inline extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java:102]: spaces will work as is on Windows and macOS?

File: extensions/grpc/stubs/pom.xml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    <artifactId>quarkus-grpc-parent</artifactId>
+    <groupId>io.quarkus</groupId>
+    <version>999-SNAPSHOT</version>
+    <relativePath>../</relativePath>
+  </parent>
+
+  <artifactId>quarkus-grpc-stubs</artifactId>
+  <name>Quarkus - gRPC - Stubs for health and reflection</name>
+
+  <dependencies>
```
Comment (gsmet) [inline extensions/grpc/stubs/pom.xml:16]: Same here about the 4 spaces convention.

**Review by gsmet — CHANGES_REQUESTED** [review]
Added some comments and questions.

File: extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java (hunk 3 above)
Comment (michalszynkiewicz) [reply extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java:102]: I don't know, I don't really have a way to check.
@rsvoboda reported that it doesn't work on linux

File: extensions/grpc/codegen/pom.xml (hunk 2 above)
Comment (michalszynkiewicz) [reply extensions/grpc/codegen/pom.xml:31]: If I remember correctly, the only reason was that there is a lot of them.
If you think it's okay to have them all in the BOM, I'll move them.

File: extensions/grpc/codegen/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/grpc/codegen/pom.xml:14]: Is it possible to add this to our formatter?

File: extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java (hunk 3 above)
Comment (michalszynkiewicz) [reply extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java:102]: To shed some more light on it, the paths are passed as separate arguments in the `ProcessBuilder`'s command, so they should be okay without backslashes before spaces. It's `protoc` that doesn't handle them properly on Linux. I don't know about other systems.

File: extensions/grpc/codegen/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply extensions/grpc/codegen/pom.xml]: fixed

File: extensions/grpc/codegen/pom.xml (hunk 2 above)
Comment (michalszynkiewicz) [reply extensions/grpc/codegen/pom.xml]: moved

**Comment by michalszynkiewicz** [comment]
I updated the doc by the way

File: docs/src/main/asciidoc/grpc-getting-started.adoc [hunk 5]
```diff
@@ -308,3 +254,88 @@ Then, open http://localhost:8080/hello/quarkus in a browser, and you should get
 
 Like any other Quarkus applications, you can package it with: `mvn package`.
 You can also package the application into a native executable with: `mvn package -Pnative`.
+
+== Generating Java files from proto with protobuf-maven-plugin
+
+Alternatively to using Quarkus code generation to generate stubs for `proto` files, you can also use
+`protobuf-maven-plugin` for this purpose.
+To do it, edit the `pom.xml` file to add the quarkus gRPC extension dependency and the JSR 305 (just under `<dependencies>`):
```
Comment (gsmet) [inline docs/src/main/asciidoc/grpc-getting-started.adoc:262]:
```suggestion
To do it, edit the `pom.xml` file to add the Quarkus gRPC extension dependency and the JSR 305 (just under `<dependencies>`):
```

Comment (gsmet) [reply docs/src/main/asciidoc/grpc-getting-started.adoc:262]: You're talking about the JSR 305 but it's not in the extract below?

File: extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java (hunk 3 above)
Comment (gsmet) [reply extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java:102]: Maybe @cescoffier can check on macOS if you can put together a test case?

**Review by gsmet — CHANGES_REQUESTED** [review]
Added some small comments on the doc.

File: docs/src/main/asciidoc/grpc-getting-started.adoc (hunk 5 above)
Comment (michalszynkiewicz) [reply docs/src/main/asciidoc/grpc-getting-started.adoc]: I only moved the section in the original commit, it seems it was a bit outdated.
I removed JSR305 and did some further fixes to the docs.

File: extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java (hunk 3 above)
Comment (michalszynkiewicz) [reply extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java:102]: @cescoffier checked and this version (without `\` before spaces) worked for him on Mac

**Review by gsmet — APPROVED** [review]

**Pull request merged**